config/src/lib.rs

```diff
@@ -1,11 +1,122 @@
-pub fn hello_world() {
-    println!("Hello World!!");
+use clap::{App, Arg, ArgMatches};
+use std::env;
+use std::fmt::Error;
+
+// Include all "stolen" ripgrep code in this module
+mod ripgrep_config;
+
+/// This trait defines the behavior that all configuration classes need to
+/// provide in order for the clap matcher to be generated from the config object
+pub trait ConfigDescription {
+    fn get_config(&self) -> Configuration;
+}
+
+/// This struct describes some properties that can be set for an application as well
+/// as the list of options that the program can understand
+/// These values (apart from the list of options) are only used to generate the
+/// console help message
+pub struct Configuration {
+    /// the name of the application
+    pub name: &'static str,
+    /// version of the application
+    pub version: &'static str,
+    /// a brief description of what the application does
+    pub about: &'static str,
+    /// the list of all possible command line options
+    pub options: Vec<ConfigOption>,
+    /// the name of an environment variable that can be used to specify an
+    /// additional config file
+    pub environment_config: &'static str,
 }
 
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+/// Represents an individual config option that the program can interpret
+pub struct ConfigOption {
```

Comment:
The answer to that is ... not simple :)
Clap has some form of support I think. On top of that, we will probably still need to implement some way of exposing the parsed values to the binary, we could stick conversion and validation code in there.
That might actually make more sense, because a type might not be sufficient for us in case we want to parse whether for example the orchestrator url looks like a valid url, not just a string.